### PR TITLE
new environment variable OAUTH_SAML_ATTRIBUTE_STATEMENTS_USERNAME to obtain username from SAML provider

### PR DESCRIPTION
**Pull request by flytreeleft**
Here is the code (from gitlab's [auth_hash.rb](https://github.com/gitlabhq/gitlabhq/blob/master/lib/gitlab/o_auth/auth_hash.rb)) to parse and transform the OmniAuth info hash:
```ruby
def name
  @name ||= get_info(:name) || "#{get_info(:first_name)} #{get_info(:last_name)}"
end

def username
  @username ||= username_and_email[:username].to_s
end

def email
  @email ||= username_and_email[:email].to_s
end

...

def username_and_email
  @username_and_email ||= begin
    username  = get_info(:username).presence || get_info(:nickname).presence
    email     = get_info(:email).presence

    username ||= generate_username(email) if email
    email    ||= generate_temporarily_email(username) if username

    {
        username: username,
        email:    email
    }
  end
end

# Get the first part of the email address (before @)
# In addtion in removes illegal characters
def generate_username(email)
  email.match(/^[^@]*/)[0].mb_chars.normalize(:kd).gsub(/[^\x00-\x7F]/, '').to_s
end
```

The `username` (or login name) will be obtained from `email` if it wasn't mapped to any attribute, there will be conflicts if the email contains same user name but different domain. And in most scenes, we need to obtain the `username` from the SAML provider.

So, mapping `username` to entries in the OmniAuth info hash explicitly should be necessary.

### CODE REVIEW

File: assets/runtime/functions [hunk 1]
```diff
@@ -546,14 +546,16 @@ gitlab_configure_oauth_bitbucket() {
 
 gitlab_configure_oauth_saml_attribute_statements() {
   if [[ -n ${OAUTH_SAML_ATTRIBUTE_STATEMENTS_EMAIL} && \
-        -n ${OAUTH_SAML_ATTRIBUTE_STATEMENTS_NAME} ]]; then
```
Comment (patope) [inline assets/runtime/functions]: Breaks existing configurations where USERNAME is currently undefined

File: assets/runtime/functions [hunk 2]
```diff
@@ -546,14 +546,16 @@ gitlab_configure_oauth_bitbucket() {
...
+        -n ${OAUTH_SAML_ATTRIBUTE_STATEMENTS_USERNAME} ]]; then
     echo "Configuring gitlab::oauth::saml::attribute_statements..."
     update_template ${GITLAB_CONFIG} \
+      OAUTH_SAML_ATTRIBUTE_STATEMENTS_USERNAME \
       OAUTH_SAML_ATTRIBUTE_STATEMENTS_EMAIL \
       OAUTH_SAML_ATTRIBUTE_STATEMENTS_NAME \
       OAUTH_SAML_ATTRIBUTE_STATEMENTS_FIRST_NAME \
       OAUTH_SAML_ATTRIBUTE_STATEMENTS_LAST_NAME
     # Remove undefined optional attributes
+    exec_as_git sed -i "/name: \\[''\\],/d" ${GITLAB_CONFIG}
```
Comment (patope) [inline assets/runtime/functions]: matches undefined first_name, last_name, name and username. Behaviour should be mentioned or make matcher unique.

**Comment by flytreeleft** [comment]
@patope Thanks, I have already submitted the fix code, please check the new code.

File: assets/runtime/functions [hunk 3]
```diff
@@ -545,15 +545,16 @@ gitlab_configure_oauth_bitbucket() {
...
     echo "Configuring gitlab::oauth::saml::attribute_statements..."
     update_template ${GITLAB_CONFIG} \
+      OAUTH_SAML_ATTRIBUTE_STATEMENTS_USERNAME \
       OAUTH_SAML_ATTRIBUTE_STATEMENTS_EMAIL \
       OAUTH_SAML_ATTRIBUTE_STATEMENTS_NAME \
       OAUTH_SAML_ATTRIBUTE_STATEMENTS_FIRST_NAME \
       OAUTH_SAML_ATTRIBUTE_STATEMENTS_LAST_NAME
     # Remove undefined optional attributes
+    exec_as_git sed -i "/\\<name: \\[''\\],/d" ${GITLAB_CONFIG}
     exec_as_git sed -i "/first_name: \\[''\\],/d" ${GITLAB_CONFIG}
```
Comment (patope) [inline assets/runtime/functions]: remove unset `username` here..

**Comment by patope** [comment]
LGTM

**Pull request closed without merging**